hamilton/lifecycle/caching.py

```diff
@@ -0,0 +1,1087 @@
+        graph: graph_types.HamiltonGraph,
+        inputs: Dict[str, Any],
+        overrides: Dict[str, Any],
+        execution_path: Collection[str],
+        **kwargs: Any,
+    ):
+        """Get code versions for all, and data version for top-level inputs and overrides
+        and setup state variables before the run
+        """
+        self.run_id = run_id
+        self._logs[run_id] = []
+        self.metadata_store.initialize()
+
+        self.graph = graph
+        self.resolve_behaviors(graph=graph)
+        # reset the code_versions and data versions before each execution
+        self.data_versions = {}
+        # code versions aren't expected to change for the lifetime of the Driver though
+        self.code_versions = {
+            node.name: self.version_code(node_name=node.name, node=node)
+            for node in self.graph.nodes
+            if node.name in execution_path
+        }
+
+        if inputs:
+            for node_name, value in inputs.items():
+                self._process_input(node_name, value)
+
+        if overrides:
+            for node_name, value in overrides.items():
+                self._process_override(node_name, value)
+
+    def run_before_node_execution(self, *args, **kwargs):
+        """required to subclass ``NodeExecutionHook``"""
+
+    # subclass `BaseDoNodeExecute` instead of `NodeExecutionMethod` to access `node.Node` object
+    def do_node_execute(
+        self,
+        *,
+        node_: hamilton.node.Node,
+        kwargs: Dict[str, Any],
+        task_id: Optional[str] = None,
+        **future_kwargs,
+    ):
+        """If the metadata store includes the context key, else compute result.
+        If a fingerprint is retrieved, but reading from cache fails, invalidate the metadata store
+        and force recompute and storage of fingerprint + result.
+        """
+        # gather the attributes typically available through NodeExecutionMethod
+        node_name = node_.name
+        node_callable = node_.callable
+        node_kwargs = kwargs
+
+        if node_.node_role == hamilton.node.NodeType.EXPAND:
+            self._expand_nodes.add(node_name)
+
+        elif node_.node_role == hamilton.node.NodeType.COLLECT:
+            self._collect_nodes.add(node_name)
+            collected_name = node_.collect_dependency
+            collected_values = node_kwargs[node_.collect_dependency]
+            # Counter is the right data structure: it's unordered, arbitrary length, allow duplicates
+            data_version_counter = collections.Counter(collected_values)
+            collected_data_version = fingerprinting.hash_mapping(dict(data_version_counter))
+            # TODO apply the right caching behavior; usually applied in `run_after_node_execution`
+            self.set_memory_metadata(
+                node_name=collected_name, task_id=task_id, data_version=collected_data_version
+            )
+
+        # node "inside" a parallel block
+        elif task_id != node_name:
```

Comment:
OK, so I feel like we can do a better job with this type of stuff. Will likely be a TODO for later, but checking equality is a non-transparent way of figuring out what we want... E.G. we should have the task object and ask it if it's inside a grouped block.

Add a TODO with caveats then we can take a look at what will be involved in wiritg it through, should be trivial.